### DISCUSSION TITLE
What do unit/integration tests for django-q users look like?

### DISCUSSION THREAD

**Discussion by MichaelAkvo** [django-q2/django-q2#74] [Q&A]
It doesn't feel like it's easy to write tests when using this library due to the tight coupling, but I may be wrong. Before I open an issue or a PR, I'd like to know how others do it.

Basically, there are a few things I'd like to test when using this library:

 - the expected task(s) has been queued
 - the arguments for the task are correct (in case of queuing multiple tasks)
 - when the task is executed that it returns the expected results

For now I had to recreate some of the code when using `async_task`

```py
class TaskTestCase(TestCase):

    def test_my_task():
        # Call method that queues the task 
        my_arg = "my arg"
        queue_my_task("my arg")

        # Check that the task was enqueued with django-q
        enqueued_task = OrmQ.objects.first()
        self.assertIsNotNone(enqueued_task)
        task_dict = SignedPackage.loads(enqueued_task.payload)
        self.assertEquals(task_dict.get("name"), "my-task")

        # And with the correct args
        task_args = task_dict.get("args")
        self.assertEquals(args[0], my_arg)

        # Direct call of function without django-q
        # This could of course be in separate unit test
        result = my_task(*task_args)
        self.assertEquals(result, "expected result")
```

As you can see, one has to know about the SignedPackage class, OrmQ object, and have intimate knowledge of what a task_dict looks like. I didn't work through how django-q actually calls the task function, nor what other side-effects there are. If the internal implementation changes somehow, the user will have to modify all their tests.

**Comment by GDay** [comment]
What I generally do is set `Sync`  to `True` and then just let the task run and check on the result of the task. That way I know that both the trigger worked and the task itself worked. For unit testing the triggering of the task, that gets a lot harder. Same for testing the scheduling.

You are right about the current way of doing that and it's indeed pretty cumbersome and prone to breaking updates.

I am currently working on a full rewrite of the library to restructure everything (more (data)classes and more event driven), which should also make testing this kind of stuff easier.
